### ISSUE TITLE
intracellular electrodes table

### ISSUE THREAD

**Issue by bendichter** [NeurodataWithoutBorders/nwb-schema#417]
In the last NWB sync, @ajtritt, @oruebel, @rly and Pam (what's her tag?) discussed creating an intracellular electrodes table. This could be used to address #416 with an optional column for cell_id, and #415 with optional columns for common solution components of the electrode internal solution. We were a bit hung up on how this would map onto our current `IntracellularElectrode` and `PatchClampSeries` ndtypes. We would like `PatchClampSeries` to be analogous to `ElectricalSeries`, where the `electrodes` field is a `DynamicTableRegion` of the electrodes table, but we already have that `PatchClampSeries.electrode` references an `IntracellularElectrode`. Possible solutions proposed:
1. Allow `PatchClampSeries` to refer to a source electrode by a link to `IntracellularElectrode` (deprecated) or by a region reference to the new table (encouraged). This would be backwards compatible, and would allow for the ideal way we would like to store things, but would create ambiguity in the way these relationships are specified.
2. Keep `PatchClampSeries` the same and have the intracellular_electrodes table have a required column with links to `IntracellularElectrode`. This would be backwards compatible and would only create a single way to link these things together, but is more complex than is really necessary and is not quite analogous to the current extracellular electrodes table.
3. The backwards incompatible way- get rid of the `IntracellularElectrode` ndtype altogether and specify all such information and links in a table, making the structure entirely analogous to the ecephys side. Making a backwards-incompatible change might cause some waves, but if we are going to do it now is the time.

**Comment by t-b** [comment]
Why do we need to touch the electrode info? Assuming https://github.com/oruebel/ndx-icephys-meta will get mainlined, you can store all the metadata you want on a per TimeSeries basis.

3. is not a solution as breaking backwards-compatiblity is a no-go.

**Comment by oruebel** [comment]
Thinking about it option ``2.`` is effectively already covered by https://github.com/oruebel/ndx-icephys-meta once it is merged. With https://github.com/oruebel/ndx-icephys-meta/issues/31 we are creating separate tables for ``stimulus`` and ``response`` and the ``intracellular_recordings`` table already has a link to the ``IntracellularElectrode``, i.e., all metadata specific to the electrode could just go directly into the ``intracellular_recordings`` table. I think this should be a reasonable option.

> Why do we need to touch the electrode info?

The main question is: (a) where should additional metadata about electrodes go? , (b) how can we keep that metadata in one place, (c) how can we keep the look-up simple, and (d) can we improve consistency between ecephys and icephys. With the current extension we can address the first issue, but it has challenges with: (b) part of the metadata would be ``IntracellularElectrode`` and other parts in the table, (c) look up of all metadata from the PatchClampSeries would require at least an additional search in the table, (d) having a table rather than an ``IntracellularElectrode`` would be more consistent with ecephys. I think we can address (c) mostly in the API by providing appropriate functions that facilitate the look-up, but (b) and (d) we cannot. However, it seems like addressing (b) and (d) would require replacing ``IntracellularElectrode`` and lead to backward incompatibilities. I.e., I think option (2) as implemented by the current extension (i.e., once https://github.com/oruebel/ndx-icephys-meta/issues/31  has been implemented) seems like a reasonable path forward. Thoughts?

**Comment by bendichter** [comment]
> all metadata specific to the electrode could just go directly into the intracellular_recordings table.

It could, but that might be pretty denormalized. This would be electrodes-specific information, and the `intracellular_recordings` table has sweep-specific information, so we would be repeating information for every sweep of a given electrode. That might not be a big deal- normalization tends to be moot if you activate compression. Also, I suppose it is possible that this would change by sweep, but I think that is unlikely. @t-b, what is your experience with this type of information? Can we safely assume things like cell and electrode solution are constant through a session?
